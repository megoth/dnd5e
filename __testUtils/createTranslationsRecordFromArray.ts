export default function createTranslationsRecordFromArray(
  url,
  list: string[]
): Record<string, string> {
  return list.reduce((memo: Record<string, string>, key) => {
    // eslint-disable-next-line no-param-reassign
    memo[`${url}#${key}`] = key;
    return memo;
  }, {});
}
