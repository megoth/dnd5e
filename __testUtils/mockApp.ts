import { AppModel, createApp, getBundleKey } from "../src/models/app";
import mockFluentBundle from "./mockFluentBundle";
import { appVocabURL } from "./mockAppIndexDataset";
import mockLanguage, { defaultLocale } from "./mockLanguage";
import mockResourceBundleMap from "./mockResourceBundleMap";
import mockAppIndex from "./mockAppIndex";

export const errorsURL = "https://example.com/global/errors.ttl";
export const faqsURL = "https://example.com/global/faqs.ttl";
export const translationsURL = "https://example.com/global/translations.ttl";
export const localizationsURL =
  "https://example.com/global/translations.en-US.ttl";

export function mockAppPropValue(val, bundleName = "global", locale = "en-US") {
  return bundleName
    ? {
        [getBundleKey(locale, bundleName)]: val,
      }
    : {
        [locale]: val,
      };
}

export default function mockApp(overrides: Partial<AppModel> = {}): AppModel {
  const currentLocale = overrides.currentLocale || defaultLocale;
  return {
    appVocabURL: overrides.appVocabURL || appVocabURL,
    currentLocale,
    fluentBundles: {
      [currentLocale]: mockFluentBundle({}, translationsURL, {
        locale: currentLocale,
      }),
      ...overrides.fluentBundles,
    },
    languages: [mockLanguage(currentLocale), ...(overrides.languages || [])],
    resourceBundles: {
      ...mockResourceBundleMap({
        label: "admin",
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...mockResourceBundleMap({
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...overrides.resourceBundles,
    },
  };
}

export function mockNorwegianApp(overrides: Partial<AppModel> = {}): AppModel {
  const locale = "nb-NO";
  return mockApp({
    currentLocale: locale,
    fluentBundles: {
      [locale]: mockFluentBundle({}, translationsURL, {
        locale,
      }),
    },
    languages: [
      mockLanguage(locale, {
        languageCode: "ðŸ‡³ðŸ‡´",
      }),
    ],
    resourceBundles: {
      ...mockResourceBundleMap({
        locale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
    },
    ...overrides,
  });
}

export function mockEmptyApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, null);
}

export function mockUnloadedApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, mockAppIndex());
}
