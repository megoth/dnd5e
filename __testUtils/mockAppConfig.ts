export default function mockAppConfig() {
  return {
    errorsUrl: "http://localhost:3000/data/errors.ttl",
    translationsUrl: "http://localhost:3000/data/translations.ttl",
  };
}
