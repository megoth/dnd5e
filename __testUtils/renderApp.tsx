import React from "react";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getFluentBundles } from "../src/models/language";
import { AppModel } from "../src/models/app";

export default function renderApp(
  app: AppModel,
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult {
  function rerender() {
    return render(
      <LocalizationProvider l10n={new ReactLocalization(getFluentBundles(app))}>
        {ui}
      </LocalizationProvider>,
      options
    );
  }
  const renderResult = rerender();
  return { ...renderResult, rerender };
}
