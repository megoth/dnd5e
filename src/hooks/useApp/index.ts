import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
