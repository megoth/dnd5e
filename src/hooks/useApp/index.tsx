import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
