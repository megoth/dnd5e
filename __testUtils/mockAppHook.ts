import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
