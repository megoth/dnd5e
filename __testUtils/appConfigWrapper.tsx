import React, { ReactNode } from "react";
import AppConfigProvider from "../src/contexts/appConfig";
import mockAppConfig from "./mockAppConfig";

interface Props {
  children: ReactNode;
}

export default function AppConfigWrapper({ children }: Props) {
  const { errorsUrl, translationsUrl } = mockAppConfig();
  return (
    <AppConfigProvider errorsUrl={errorsUrl} translationsUrl={translationsUrl}>
      {children}
    </AppConfigProvider>
  );
}
