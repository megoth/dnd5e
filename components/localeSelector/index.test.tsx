import React from "react";
import useApp from "../../src/hooks/useApp";
import LocaleSelector, { TESTID_LOCALE_SELECTOR_LANGUAGE } from "./index";
import mockAppHook from "../../__testUtils/mockAppHook";
import { mockNorwegianApp } from "../../__testUtils/mockApp";
import renderApp from "../../__testUtils/renderApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("LocaleSelector", () => {
  it("renders nothing when only one language is given", () => {
    const app = mockAppHook(mockedAppHook);
    const { asFragment, queryAllByTestId } = renderApp(app, <LocaleSelector />);
    expect(asFragment()).toMatchSnapshot();
    expect(queryAllByTestId(TESTID_LOCALE_SELECTOR_LANGUAGE)).toHaveLength(0);
  });

  it("renders controls to select other languages if multiple languages are given", () => {
    const app = mockAppHook(mockedAppHook, mockNorwegianApp());
    const { asFragment, queryAllByTestId } = renderApp(app, <LocaleSelector />);
    expect(asFragment()).toMatchSnapshot();
    expect(queryAllByTestId(TESTID_LOCALE_SELECTOR_LANGUAGE)).toHaveLength(1);
  });
});
