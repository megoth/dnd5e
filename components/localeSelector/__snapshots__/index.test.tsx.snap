// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleSelector renders controls to select other languages if multiple languages are given 1`] = `
<DocumentFragment>
  <span
    data-testid="locale-selector-language"
  >
    <a
      class="bg-gray-100 py-1 px-2 rounded hover:bg-gray-300 focus:outline-none focus:ring-2 focus:ring-gray-800"
      href="/?locale=%F0%9F%87%B3%F0%9F%87%B4"
    >
      ðŸ‡ºðŸ‡¸Â [Translation for https://example.com/global/translations.ttl#translateTo does not exist]
    </a>
  </span>
</DocumentFragment>
`;

exports[`LocaleSelector renders nothing when only one language is given 1`] = `<DocumentFragment />`;
